sygnal/webpushpushkin.py

```diff
@@ -0,0 +1,167 @@
+# limitations under the License.
+import json
+import logging
+import time
+from io import BytesIO
+from json import JSONDecodeError
+
+from pywebpush import webpush, WebPushException
+from py_vapid import Vapid
+from opentracing import logs, tags
+from prometheus_client import Counter, Gauge, Histogram
+from twisted.enterprise.adbapi import ConnectionPool
+from twisted.internet.defer import DeferredSemaphore
+from twisted.web.client import FileBodyProducer, HTTPConnectionPool, readBody
+from twisted.web.http_headers import Headers
+
+from sygnal.exceptions import (
+    NotificationDispatchException,
+    TemporaryNotificationDispatchException,
+)
+from sygnal.helper.context_factory import ClientTLSOptionsFactory
+from sygnal.helper.proxy.proxyagent_twisted import ProxyAgent
+from sygnal.utils import NotificationLoggerAdapter, twisted_sleep
+
+from .exceptions import PushkinSetupException
+from .notifications import ConcurrencyLimitedPushkin
+
+logger = logging.getLogger(__name__)
+
+MAX_TRIES = 3
+RETRY_DELAY_BASE = 10
+MAX_BYTES_PER_FIELD = 1024
+
+DEFAULT_MAX_CONNECTIONS = 20
+
+
+class WebpushPushkin(ConcurrencyLimitedPushkin):
+    """
+    Pushkin that relays notifications to Google/Firebase Cloud Messaging.
+    """
+
+    UNDERSTOOD_CONFIG_FIELDS = {
+        "type",
+        "max_connections",
+        "vapid_private_key",
+        "vapid_contact_email",
+    } | ConcurrencyLimitedPushkin.UNDERSTOOD_CONFIG_FIELDS
+
+    def __init__(self, name, sygnal, config):
+        super(WebpushPushkin, self).__init__(name, sygnal, config)
+
+        nonunderstood = set(self.cfg.keys()).difference(self.UNDERSTOOD_CONFIG_FIELDS)
+        if len(nonunderstood) > 0:
+            logger.warning(
+                "The following configuration fields are not understood: %s",
+                nonunderstood,
+            )
+
+        self.http_pool = HTTPConnectionPool(reactor=sygnal.reactor)
+        self.max_connections = self.get_config(
+            "max_connections", DEFAULT_MAX_CONNECTIONS
+        )
+        self.connection_semaphore = DeferredSemaphore(self.max_connections)
+        self.http_pool.maxPersistentPerHost = self.max_connections
+
+        tls_client_options_factory = ClientTLSOptionsFactory()
+
+        # use the Sygnal global proxy configuration
+        proxy_url = sygnal.config.get("proxy")
+
+        self.http_agent = ProxyAgent(
+            reactor=sygnal.reactor,
+            pool=self.http_pool,
+            contextFactory=tls_client_options_factory,
+            proxy_url_str=proxy_url,
+        )
+        self.http_agent_wrapper = HttpAgentWrapper(self.http_agent)
+
+        privkey_filename = self.get_config("vapid_private_key")
+        self.vapid_private_key = Vapid.from_file(private_key_file=self.get_config("vapid_private_key"))
+        vapid_contact_email = self.get_config("vapid_contact_email")
+        self.vapid_claims = {"sub": "mailto:{}".format(vapid_contact_email)}
```

Comment:
Will this blow up if the VAPID configuration isn't present? Should we handle that better?

Reply:
indeed, I assume by throwing a `PushkinSetupException` if unhappy